sdc/datatypes/hpat_pandas_series_functions.py

```diff
@@ -4930,22 +4932,43 @@ def sdc_pandas_series_combine(self, other, func, fill_value=None):
     if not isinstance(fill_value, (types.Omitted, types.NoneType, types.Number)) and fill_value is not None:
         ty_checker.raise_exc(fill_value, 'number', 'fill_value')
 
+    fill_is_default = isinstance(fill_value, (types.Omitted, types.NoneType)) or fill_value is None
+
+    sig = func.get_call_type(cpu_target.typing_context, [self.dtype, other.dtype], {})
+    ret_type = sig.return_type
+
+    fill_dtype = types.float64 if fill_is_default else fill_value
+    res_dtype = find_common_dtype_from_numpy_dtypes([], [ret_type, fill_dtype])
+
     def sdc_pandas_series_combine_impl(self, other, func, fill_value=None):
 
-        if fill_value is None:
-            fill_value = numpy.nan
+        if fill_value is not None:
+            _fill_value = fill_value
+        else:
+            _fill_value = numpy.nan
 
-        len_val = max(len(self), len(other))
-        result = numpy.empty(len_val, self._data.dtype)
-        for ind in range(len_val):
-            val_self = self._data[ind]
-            val_other = other._data[ind]
-            if len(self) < ind + 1:
-                val_self = fill_value
-            if len(other) < ind + 1:
-                val_other = fill_value
-            result[ind] = func(val_self, val_other)
+        indexes, self_indexes, other_indexes = sdc_join_series_indexes(self.index, other.index)
+        len_val = len(indexes)
 
-        return pandas.Series(result)
+        result = numpy.empty(len_val, res_dtype)
+
+        chunks = parallel_chunks(len_val)
+        for i in prange(len(chunks)):
+            chunk = chunks[i]
+            for j in range(chunk.start, chunk.stop):
+                self_idx = self_indexes[j]
+                if self_idx == -1:
+                    val_self = _fill_value
+                else:
+                    val_self = self[self_idx]._data[0]
```

Comment:
self_idx (and other_idx) is position in the Series, not the index, so instead of using getitem on a Series, that performs index lookup and returns a Series, so that you have to take _data[0] from it, you can just write:
```suggestion
                    val_self = self._data[self_idx]
```